src/Language/SystemVerilog/Parser/Preprocess.hs

```diff
@@ -148,6 +158,64 @@ isIdentChar ch =
     ('0' <= ch && ch <= '9') ||
     (ch == '_') || (ch == '$')
 
+-- Update a Cond for an `else block: if we've never had a hit, this is the one
+-- for us. Otherwise a previous block was true, so it isn't.
+elseCond :: Cond -> Cond
+elseCond NeverTrue = CurrentlyTrue
+elseCond _ = NeverTrue
+
+-- Return a Cond for whether a name is defined (assuming that it hasn't been
+-- defined previously).
+ifdefCond :: Env -> String -> Cond
+ifdefCond env name = if Map.member name env then CurrentlyTrue else NeverTrue
+
+-- Update a Cond for an `elsif block. The boolean argument is whether the
+-- `elsif block's test is true.
+elsifCond :: Env -> String -> Cond -> Cond
+elsifCond env name c =
+    case c of
+        NeverTrue -> ifdefCond env name
+        _ -> PreviouslyTrue
+
+-- Push a condition onto the top of the preprocessor condition stack
+pushCondStack :: Cond -> PPS ()
+pushCondStack c = getCondStack >>= (\ cs -> setCondStack (c : cs))
+
+-- Pop the top from the preprocessor condition stack. If the stack is empty,
+-- raise a lexical error based on name (the name of the directive that caused
+-- the problem).
+popCondStack :: String -> PPS Cond
+popCondStack name = do
+    cs <- getCondStack
+    case cs of
+        [] -> lexicalError $
+              "`" ++ name ++ " directive outside of an `if/`endif block"
+        c : cs' -> setCondStack cs' >> pure c
+
+-- Update the preprocessor condition stack for an `ifdef
+csIfdef :: Env -> String -> PPS ()
+csIfdef env name = pushCondStack (ifdefCond env name)
+
+-- Update the preprocessor condition stack for an `ifndef
+csIfndef :: Env -> String -> PPS ()
+csIfndef env name = pushCondStack (elseCond (ifdefCond env name))
```

Comment:
Ok, can do. My motivation was really that more functions means more helpful stack traces with `-xc`, but since my code is clearly bug-free (hah!), of course that won't matter...